docs/specifications/4-standard-compliance-suite.md

```diff
@@ -0,0 +1,166 @@
+can be useful when investigating the test:
+- **Introduced** specifies the very first *released* version of `jsii` that
+  included the test. This implies previous versions of `jsii` may not have
+  conformed to the test.
+- **Related Issues** lists any and all issues related to the test case. This
+  includes the issue that was created to track the introduction of the test, and
+  other issues such as bugs that prompted the introduction of a new compliance
+  test.
+- **Revised** lists all versions of `jsii` after the initial introduction of the
+  test case, for which the release included a revision of the test case. Trivial
+  revisions (such as correcting typos, or re-phrasing the description without
+  changing the tests case's semantic value) should not be included in the list.
+
+After the attributes table, a **TypeScript** block of code describes the
+canonical form of the test. It includes any type declaration that is used by the
+test (so the code example is self-contained). Assertions performed by the test
+should be written in the form of [`jest`] expectations.
+
+[`jest`]: https://jestjs.io/docs/en/getting-started
+
+Finally, another code block details the sequence of messages that should be
+exchanged between the *host* and `node` processes during the execution of the
+test case, such that implementations can assert conherent behavior. The dialogue
+is the sequence of JSON formatted messages, from the perspective of the *host*
+app, using the following notation:
+* Messages sent by the *host* runtime to the `node` process:
+  ```
+  > { "api": "foo" }
+  ```
+* Messages received by the *host* runtime from the `node` process:
+  ```
+  < { "result": "bar" }
+  ```
+* Comments to improve readability of the trace:
+  ```
+  # Comment continues until the end of the line
+  ```
+* Blank lines can be added to logically group trace elements
+
+> :question: is there a need to support some form of a capture mechanism to
+> provision for non-deterministric results, or non-normative elements such as
+> the exact Object IDs issued for created instances?
+
+#### Template
+
+Below is the template markdown to copy-paste when introducing a new test case in
+the compliance suite. New tests should always be added at the very end of the
+category they belong to, right after the last test in said category.
+
+````md
+### Test Category
+#### Test Case Name
+
+A short english language description of what property this test verifies. The
+description should include enough detail for a reader to be able to understand
+the test without having to search for any additional information. Prefer a long,
+unambiguous description to a terse one that could be subject to interpretation.
+
+Attribute      | Value
+---------------|----------------------------------------------------------------
+Introduced     | 1.2.3
+Related Issues | #123, aws/aws-cdk#456
+
+##### Reference Implementation
+```ts
```

Comment:
This is not entirely clear to me. Looks like this consists of 2 parts:

* the parts that have to be present in TypeScript-land for the test to run (loaded into the kernel)
*  an assertion on the behavior of that code as it is being exercised, which needs to be rewritten into every client language.

Did I get that right?

I imagine there will be a lot of duplication between the kernel code of various tests. Is there going to be a single set of classes that cover everything, or are new classes going to be written for every individual test?

Reply:
That's correct - basically there needs to be:
- Type entities defined in TypeScript
- Expectation/assertions that need writing in each language.

The may end up being a lot of duplication, but I would rather keep the tests self-contained as much as possible (open to change if that turns out very inconvenient). The idea would be to code-generate the combination of all things in per-test namespaces, automatically.

Reply:
I was also really curious to see how this works with each binding language – what's the experience to implement the tests in those languages?